server/src/main/java/org/opensearch/indices/replication/SegmentReplicationSourceService.java

```diff
@@ -0,0 +1,153 @@
+
+        transportService.registerRequestHandler(
+            Actions.GET_CHECKPOINT_INFO,
+            ThreadPool.Names.GENERIC,
+            CheckpointInfoRequest::new,
+            new CheckpointInfoRequestHandler()
+        );
+        transportService.registerRequestHandler(
+            Actions.GET_SEGMENT_FILES,
+            ThreadPool.Names.GENERIC,
+            GetSegmentFilesRequest::new,
+            new GetSegmentFilesRequestHandler()
+        );
+    }
+
+    private class CheckpointInfoRequestHandler implements TransportRequestHandler<CheckpointInfoRequest> {
+        @Override
+        public void messageReceived(CheckpointInfoRequest request, TransportChannel channel, Task task) throws Exception {
+            final ReplicationCheckpoint checkpoint = request.getCheckpoint();
+            logger.trace("Received request for checkpoint {}", checkpoint);
+            final CopyState copyState = getCachedCopyState(checkpoint);
+            channel.sendResponse(
+                new CheckpointInfoResponse(
+                    copyState.getCheckpoint(),
+                    copyState.getMetadataSnapshot(),
+                    copyState.getInfosBytes(),
+                    copyState.getPendingDeleteFiles()
+                )
+            );
+        }
+    }
+
+    class GetSegmentFilesRequestHandler implements TransportRequestHandler<GetSegmentFilesRequest> {
+        @Override
+        public void messageReceived(GetSegmentFilesRequest request, TransportChannel channel, Task task) throws Exception {
+            if (isInCopyStateMap(request.getCheckpoint())) {
+                // TODO send files
+            } else {
+                channel.sendResponse(TransportResponse.Empty.INSTANCE);
+            }
+        }
+    }
+
+    /**
+     * Operations on the {@link #copyStateMap} member.
+     */
+
+    /**
+     * A synchronized method that checks {@link #copyStateMap} for the given {@link ReplicationCheckpoint} key
+     * and returns the cached value if one is present. If the key is not present, a {@link CopyState}
+     * object is constructed and stored in the map before being returned.
+     */
+    private synchronized CopyState getCachedCopyState(ReplicationCheckpoint checkpoint) throws IOException {
+        if (isInCopyStateMap(checkpoint)) {
+            final CopyState copyState = fetchFromCopyStateMap(checkpoint);
+            copyState.incRef();
+            return copyState;
+        } else {
+            // From the checkpoint's shard ID, fetch the IndexShard
+            ShardId shardId = checkpoint.getShardId();
+            final IndexService indexService = indicesService.indexService(shardId.getIndex());
+            final IndexShard indexShard = indexService.getShard(shardId.id());
+            // build the CopyState object and cache it before returning
+            final CopyState copyState = new CopyState(indexShard);
+            // TODO This will add with the latest checkpoint, not the one from the request
```

Comment:
I think this should be the checkpoint requested, not the one returned.  This way replicas requesting the same checkpoint will copy to the same returned checkpoint in the CopyState.  We'll be storing the latest received checkpoint on replicas to replay after a copy event completes.